platform/FileHandle.h

```diff
@@ -152,8 +156,88 @@ class FileHandle {
+     *  The default is blocking.
+     *
+     *  @param blocking true for blocking mode, false for non-blocking mode.
+     */
+    virtual int set_blocking(bool blocking) {
+        return -1;
+    }
+
+    /** Check for poll event flags
+     * The input parameter can be used or ignored - the could always return all events,
+     * or could check just the events listed in events.
+     * Call is non-blocking - returns instantaneous state of events.
+     * Whenever an event occurs, the derived class must call _poll_change().
+     * @param events bitmask of poll events we're interested in - POLLIN/POLLOUT etc.
+     * @return bitmask of poll events that have occurred.
+     */
+    virtual short poll(short events) const {
+        // Possible default for real files
+        return POLLIN | POLLOUT;
+    }
+    /** Returns true if the FileHandle is writable.
+     *  Definition depends upon the subclass implementing FileHandle.
+     *  For example, if the FileHandle is of type Stream, writable() could return
+     *  true when there is ample buffer space available for write() calls.
+     */
+    bool writable() const
+    {
+        return poll(POLLOUT) & POLLOUT;
+    }
+
+    /** Returns true if the FileHandle is readable.
+     *  Definition depends upon the subclass implementing FileHandle.
+     *  For example, if the FileHandle is of type Stream, readable() could return
+     *  true when there is something available to read.
+     */
+    bool readable() const
+    {
+        return poll(POLLIN) & POLLIN;
+    }
+
+    /** Register a callback on state change of the file.
+     *
+     *  The specified callback will be called on state changes such as when
+     *  the file can be written to or read from.
+     *
+     *  The callback may be called in an interrupt context and should not
+     *  perform expensive operations.
+     *
+     *  Note! This is not intended as an attach-like asynchronous api, but rather
+     *  as a building block for constructing  such functionality.
+     *
+     *  The exact timing of when the registered function
+     *  is called is not guaranteed and susceptible to change. It should be used
+     *  as a cue to make read/write/poll calls to find the current state.
+     *
+     *  @param func     Function to call on state change
+     */
+    void sigio(Callback<void()> func);
+
+    /** Issue sigio to user - used by mbed::_poll_change */
+    void _send_sigio()
+    {
+        if (_callback) {
+            _callback();
+        }
+    }
+
+private:
+    Callback<void()> _callback;
 };
 
+/** Not a member function
+ *  This call is equivalent to posix fdopen().
+ *  Returns a pointer to std::FILE
+ *  It associates a Stream to an already opened file descriptor (FileHandle)
+ *  @param fh, a pointer to an opened file descriptor
+ *  @param mode, operation upon the file descriptor, e.g., 'wb+'*/
+
+std::FILE *fdopen(FileHandle *fh, const char *mode);
```

Comment:
It's in retarget.cpp. Personally, I'm not a fan of retarget.h being something for apps to include, so FileHandle.h seems reasonable to me. fdopen(int) would be stdio.h - fdopen is primarily a function about stdio.

This particular overload of it for FileHandle seems reasonable for FileHandle.h - with a bit of restructuring I can imagine the implementation moving to FileHandle.cpp, calling some other API in retarget.cpp but with the current implementation of the "POSIX" bits, it ends up going there directly.

The main reason for putting the current implementation in retarget.cpp is that the implementation requires knowledge of the "@xxxxxxxx". But that's an implementation detail, nothing to do with the API.